cpp/src/column/column.cu

```diff
@@ -275,6 +278,93 @@ struct create_column_from_view_vector {
 
 };
 
+struct concatenate_using_partition_map {
+  std::vector<column_view> const& views;
+  cudaStream_t stream;
+  rmm::mr::device_memory_resource* mr;
+
+  template <typename T,
+      std::enable_if_t<is_fixed_width<T>()>* = nullptr>
+  std::unique_ptr<column> operator()() {
+
+    // Compute the partition offsets
+    thrust::host_vector<size_type> offsets(views.size() + 1);
+    thrust::transform_inclusive_scan(thrust::host,
+      views.begin(), views.end(), offsets.begin() + 1,
+      [](column_view const& col) {
+        return col.size();
+      },
+      thrust::plus<size_type>{});
+    thrust::device_vector<size_type> d_offsets(offsets);
+    auto const output_size = offsets.back();
+
+    // Compute partition map
+    // NOTE instead of scattering `i` and doing a "max scan", scatter 1s
+    // for all but the first offset and do an inclusive scan
+    thrust::device_vector<size_type> d_partition_map(output_size);
+    auto const const_it = thrust::make_constant_iterator(size_type{1});
+    thrust::scatter(rmm::exec_policy(stream)->on(stream),
+      const_it, const_it + (d_offsets.size() - 2), // skip first and last offset
+      d_offsets.begin() + 1, // skip first offset, leaving it set to 0
+      d_partition_map.begin());
+    thrust::inclusive_scan(rmm::exec_policy(stream)->on(stream),
+      d_partition_map.begin(), d_partition_map.end(),
+      d_partition_map.begin());
+
+    // Transform views to array of data pointers
+    thrust::host_vector<T const*> data_ptrs(views.size());
+    std::transform(views.begin(), views.end(), data_ptrs.begin(),
+      [](column_view const& col) {
+        return col.data<T>();
+      });
+    thrust::device_vector<T const*> d_data_ptrs(data_ptrs);
+
+    // TODO add null mask support
+    auto out_col = experimental::detail::allocate_like(views.front(), output_size,
+      experimental::mask_allocation_policy::NEVER, mr, stream);
+    auto out_view = out_col->mutable_view();
+
+    // Initialize each output row from its corresponding input view
+    auto const* offsets_ptr = d_offsets.data().get();
+    auto const* partition_ptr = d_partition_map.data().get();
+    auto const** data_ptr = d_data_ptrs.data().get();
+    thrust::tabulate(rmm::exec_policy(stream)->on(stream),
+      out_view.begin<T>(), out_view.end<T>(),
+      [offsets_ptr, partition_ptr, data_ptr] __device__ (auto i) {
```

Comment:
Just a heads up that you can do this in the capture and avoid having to declare the variables above. 
```suggestion
      [offsets_ptr = d_offsets.data().get(), partition_ptr = d_partition_map.data().get(), data_ptr = d_data_ptrs.data().get()] __device__ (auto i) {
```